rust/apps/zcash/src/pczt/parse.rs

```diff
@@ -0,0 +1,573 @@
+                    "enc_ciphertext field for Orchard action is undecryptable".into(),
+                )),
+                // We couldn't directly decrypt a zero-valued note. This is okay because
+                // it is checked elsewhere that the direct details in the PCZT are valid,
+                // and thus the output definitely has zero value and thus can't adversely
+                // affect balance. The decryption failure means that the `enc_ciphertext`
+                // contains no in-band data (as is the case for e.g. dummy outputs).
+                (None, _) => Ok(None),
+            },
+        };
+
+    let mut keys = vec![(Some(external_ovk), false), (Some(internal_ovk), true)];
+
+    if let Some(ovk) = transparent_internal_ovk {
+        keys.push((Some(ovk), true));
+    }
+
+    // Require that we can view all non-zero-valued outputs by falling back on direct
+    // decryption.
+    keys.push((None, false));
+
+    let mut parsed_to = None;
+
+    for key in keys {
+        // TODO: Should this be a soft error ("catch" the decryption failure error here
+        // and store it in `ParsedTo` to inform the user that an output of their
+        // transaction is unreadable, but still give them the option to sign), or a hard
+        // error (throw via `?` and cause the entire transaction to be unsignable)? The
+        // answer will likely depend on what checks are performed on the PCZT prior to
+        // here. For now we've chosen the latter (hard error).
+        let output = decode_output(key.0, key.1)?;
+        match output {
+            Some(output) => {
+                parsed_to = Some(output);
+                break;
+            }
+            None => continue,
+        }
+    }
+
+    match parsed_to {
+        None => {
+            let (address, is_dummy) = match (output.user_address(), value) {
+                (Some(addr), _) => Ok((addr.clone(), false)),
+                (None, 0) => Ok(("Dummy output".into(), true)),
+                (None, _) => Err(ZcashError::InvalidPczt(
+                    "missing user address for Orchard output".into(),
+                )),
+            }?;
+            Ok(ParsedTo::new(
+                address,
+                format_zec_value(value as f64),
+                value,
+                false,
+                is_dummy,
+                None,
+            ))
+        }
+        Some(x) => Ok(x),
+    }
+}
+
+fn decode_memo(memo_bytes: [u8; 512]) -> Option<String> {
+    let first = memo_bytes[0];
+
+    //decode as utf8.
+    if first <= 0xF4 {
+        let mut temp_memo = memo_bytes.to_vec();
+        temp_memo.reverse();
+        let mut result = vec![];
+        temp_memo.iter().for_each(|v| {
+            if *v != 0 {
+                result.push(*v);
+            }
+        });
+        result.reverse();
+
+        return Some(String::from_utf8(result).unwrap());
+    }
```

Comment:
This is subtly incorrect: rather than only stripping trailing `nul` bytes from the memo bytes, it is stripping _every_ `nul` byte from the memo bytes. `nul` is a valid byte in UTF-8 (as part of the NUL character).

If a NUL character cannot be included in rendered memo strings (if Keystone requires C-strings), then it should be replaced with an appropriate Unicode replacement character.

Reply:
I aim to remove until the first non-zero value but implement a wrong one.